### PR TITLE
Fix planning. Previously last_run_time was not being used correctly

### PR DESCRIPTION
**Pull request by oschwald**
This could cause steps to be incorrectly built or not built. 

Also, eliminated multiprocess choke points where we waited for a step group to finish before doing anything more.

Removed deprecated `Stepford::Planner` class.

### CODE REVIEW

File: lib/Stepford/Graph.pm [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+    return \%args;
+}
+
+# Note: this is intentionally depth-first traversal
+sub traverse {
+    my $self = shift;
+    my $cb   = shift;
+
+    $_->traverse($cb) for @{ $self->_children_graphs };
```
Comment (eilara) [inline lib/Stepford/Graph.pm]: Could we get to a deep recursion warning here? Can't see a trivial way to convert to iteration, though there probably is, but I can vaguely picture how you would convert it to goto &sub + continuation-passing-style to get it tail-recursive.

Or maybe it is not a problem, especially as there are way to increase the Perl deep recursion limit?

Comment (oschwald) [inline lib/Stepford/Graph.pm]: Yes, the warning is possible, but the graph would have to be 100 steps deep. I think our biggest graphs max out at a depth of 5 or 6 steps. I agree that tail recursion could make sense, but I also doubt our graphs will ever get that deep.

File: lib/Stepford/Graph.pm [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+# This checks whether we should run the step. It is meant to be run after we
+# determine that the step is in a state where it _can_ run. This primarily
+# looks at the internal state of the step, e.g., last run times.
+#
+# This is generally only called once, immediately before we run the step or
+# decide never to run it.
+#
+# can and should are separated as they serve two different purposes in the
+# Runner's flow control.
```
Comment (eilara) [inline lib/Stepford/Graph.pm]: Confused me with _should_run_step method of runner, maybe this method (step is up to date) should be renamed should_run_step?

File: lib/Stepford/Graph.pm [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+sub traverse {
+    my $self = shift;
+    my $cb   = shift;
+
+    $_->traverse($cb) for @{ $self->_children_graphs };
+    $cb->($self);
+    return;
+}
+
+# This checks is the step/graph is in a state where we can run it, e.g.,
```
Comment (eilara) [inline lib/Stepford/Graph.pm]: ..if the step/graph is in...
or
..is the step/graph in a...

File: lib/Stepford/Graph.pm [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+        $args{$dep} = $productions{$dep};
+    }
+
+    $args{logger} = $self->logger;
+
+    return \%args;
+}
+
+# Note: this is intentionally depth-first traversal
```
Comment (eilara) [inline lib/Stepford/Graph.pm]: maybe add: as required for a topological sort

File: lib/Stepford/Graph.pm [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    required => 1,
+);
+
+has logger => (
+    is       => 'ro',
+    isa      => Logger,
+    required => 1,
+);
+
+has step => (
```
Comment (eilara) [inline lib/Stepford/Graph.pm]: Confused me- maybe better step_class? Anyways, the ISA constraint is wrong but strangely working

File: lib/Stepford/Graph.pm (hunk 2 above)
Comment (oschwald) [inline lib/Stepford/Graph.pm]: I went back and for on this one initially, as it used to be called `should_run_step`. However, I was concerned that this creates confusing with `can_run_step` and it doesn't describe what it is doing.  That said, I am happy to change it back if you think `should_run_step` is clearer.

The method in the runner checks both this and `can_run_step`.

File: lib/Stepford/Graph.pm (hunk 5 above)
Comment (oschwald) [inline lib/Stepford/Graph.pm]: I'll rename it. The `isa` for this predates my changes. It was used for classes previously. See, e.g., https://github.com/maxmind/Stepford/blob/master/lib/Stepford/Runner.pm#L67.

**Comment by oschwald** [comment]
I made the changes besides the `should_run_step` one. I am open to changing that too, but I am not sure what the best change is.

Oh, I left the `isa` as that seems to be the best constraint available.

**Pull request merged**